File: app/javascript/channels/hub/notification_channel.js
import consumer from "channels/consumer"
import { NotificationAPI, NotificationDOM } from "utilities/notification_utils"
import notificationEventBus from "utilities/event_bus"
import { NotificationType } from "utilities/notification_types"

/**
 * Notification Channel for real-time notifications via Solid Cable
 * Handles receiving notifications, displaying toasts, and updating counters
 */

// Log channel connection status for debugging
const DEBUG = false;
const log = (...args) => DEBUG && console.log('[NotificationChannel]', ...args);

// Safely create notification channel
// Export as a global for components that need direct access
let notificationChannel = null;

try {
  // Check if consumer is available
  if (consumer) {
    notificationChannel = consumer.subscriptions.create("Hub::NotificationChannel", {
      connected() {
        log('Connected to notification channel')
        this.sendPendingActions()
      },

      disconnected() {
        log('Disconnected from notification channel')
      },
      
      // Store pending actions when disconnected
      pendingActions: [],
      
      // Send any pending actions when we reconnect
      sendPendingActions() {
        if (this.pendingActions.length > 0) {
          log('Sending pending actions:', this.pendingActions.length)
          
          this.pendingActions.forEach(action => {
            this.perform(action.name, action.data)
          })
          
          this.pendingActions = []
        }
      },
      
      // Safely perform an action or queue it if disconnected
      safePerform(actionName, data = {}) {
        // Check if connection is in OPEN state (1)
        if (this.consumer && this.consumer.connection && this.consumer.connection.webSocket && this.consumer.connection.webSocket.readyState === 1) {
          return this.perform(actionName, data)
        } else {
          log('Connection not available, queuing action:', actionName)
          this.pendingActions.push({ name: actionName, data })
          return false
        }
      },
      
      // Request the current unread count from the server
      // This is useful when we need to refresh the count without polling
      requestCount() {
        log('Requesting unread count from server')
        return this.safePerform('request_count')
      },

      received(data) {
        log('Received message:', data)
        
        // Skip if no data received
        if (!data) {
          console.error('[NotificationChannel] Received empty data')
          return
        }
        
        // Handle different types of messages from the server
        if (data.type || data.action) {
          // Get the message type (support both new type and legacy action)
          const messageType = data.type || data.action
          
          switch (messageType) {
            case 'notification_dismissed':
              this.handleNotificationDismissed(data)
              break
              
            case 'unread_count_updated':
              this.handleUnreadCountUpdated(data)
              break
              
            case 'all_notifications_read':
              // All notifications were marked as read
              notificationEventBus.emit('notification:all-read')
              this.handleUnreadCountUpdated({ count: 0 })
              break
              
            case 'ping':
              // Server checking if we're alive
              this.safePerform('pong')
              break
              
            default:
              log('Unknown message type received:', messageType)
          }
        }
        // Handle notification object (most common case)
        else if (data.notification) {
          // First, update the counter if unread_count is provided
          if (data.notification.unread_count !== undefined) {
            this.handleUnreadCountUpdated({ count: data.notification.unread_count })
          } else {
            // Otherwise fetch the latest count from the server
            this.updateNotificationCount()
          }

          // Then display the toast notification if requested
          const showToast = data.notification.show_toast === true
          if (showToast) {
            this.showToastNotification(data.notification)
          }
          
          // Use event bus to notify the application of the new notification
          notificationEventBus.emit('notification:received', {
            count: data.notification.unread_count || 1,
            notification: data.notification
          })
        } else {
          log('Received unrecognized message format:', data)
        }
      },
      
      // Handle notification dismissed action
      handleNotificationDismissed(data) {
        if (data.id) {
          log('Notification dismissed:', data.id)
          // Emit event with the correct data structure
          notificationEventBus.emit('notification:dismissed', { id: data.id })
          
          // Use the DOM utility to broadcast the event and update counter
          NotificationDOM.broadcastDismissEvent(data.id)
          
          // If the count was provided in the message, update it directly
          if (data.count !== undefined) {
            this.handleUnreadCountUpdated({ count: data.count })
          } else {
            // Otherwise fetch the latest count after a brief delay
            // to allow the server time to process the dismissal
            setTimeout(() => this.updateNotificationCount(), 300)
          }
        }
      },
      
      // Handle unread count updated action
      handleUnreadCountUpdated(data) {
        if (data.count !== undefined) {
          log('Unread count updated:', data.count)
          // Use event bus to notify all components of count change
          notificationEventBus.emit('notification:count-updated', { count: data.count })
          
          // Also update DOM directly for older components
          NotificationDOM.updateCounter(data.count)
        }
      },
      
      // Mark notification as read
      markAsRead(notificationId) {
        log('Marking notification as read:', notificationId)
        return this.safePerform('mark_as_read', { id: notificationId })
      },
      
      // Dismiss a notification
      dismissNotification(notificationId) {
        log('Dismissing notification:', notificationId)
        return this.safePerform('dismiss', { id: notificationId })  
      },
      
      // Mark notification as displayed
      markAsDisplayed(notificationId) {
        log('Marking notification as displayed:', notificationId)
        return this.safePerform('mark_as_displayed', { id: notificationId })
      },

      // Show a toast notification
      showToastNotification(notification) {
        // First, try to find the toast container by ID
        let toastContainer = document.getElementById('toast-container')
        
        // If not found by ID, try to find it by data attribute
        if (!toastContainer) {
          toastContainer = document.querySelector('[data-toasts-target="container"]')
        }
        
        // If not found by data attribute, try to find it by controller
        if (!toastContainer) {
          toastContainer = document.querySelector('[data-controller="toasts"]')
        }
        
        if (toastContainer) {
          // Make sure notification has required fields for toast
          const toastNotification = {
            id: notification.id,
            message: notification.message,
            notification_type: notification.notification_type,
            created_at: notification.created_at,
            url: notification.url
          }
          
          // Create and dispatch the event
          const event = new CustomEvent('notification:toast', {
            detail: toastNotification,
            bubbles: true // Make sure event bubbles up through the DOM
          })
          
          // Dispatch both to document and directly to container for maximum compatibility
          document.dispatchEvent(event)
          toastContainer.dispatchEvent(event)
        }
      },

      // Update the notification count in the UI
      async updateNotificationCount() {
        try {
          const data = await NotificationAPI.getCount()
          
          // Use NotificationDOM utility to update the counter
          NotificationDOM.updateCounter(data.count)
          
          // We no longer emit the event here to prevent loops
          // Direct DOM updates are preferred - components that need this data
          // should listen to the Action Cable channel directly
        } catch (error) {
          // Handle error silently
        }
      }
    })
  }
} catch (error) {
  // Handle error silently
}

// Export for module imports
export default notificationChannel

// Also expose globally for direct access from components
window.notificationChannel = notificationChannel

File: app/channels/hub/notification_channel.rb
module Hub
  # The NotificationChannel handles real-time notification delivery via Solid Cable
  # This channel streams notifications directly to the user's browser and
  # handles user-initiated actions like marking notifications as read or dismissed.
  #
  # The channel is used by the NotificationService to broadcast notifications
  # and by the JavaScript client to receive and manage notifications.
  #
  # Rails 8 Solid Cable features used:
  # - Stream for specific records (stream_for)
  # - Built-in presence tracking
  # - Enhanced security with channel subscriptions
  class NotificationChannel < ApplicationCable::Channel
    include ActionCable::Channel::Streams
    include ActionCable::Channel::Broadcasting
    
    # Declares expected params for channel subscription
    # This is a new security feature in Rails 8
    # Use params.permit or verify configuration in connection setup instead
    
    # Subscribes the current user to their personal notification stream
    # @return [void]
    def subscribed
      # Log connection state for debugging
      Rails.logger.debug { "NotificationChannel#subscribed - Connection User: #{connection.current_user&.id}, Current.user: #{Current.user&.id}" }
      
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      
      # Ensure we have a current user
      return reject_connection("No authenticated user") unless user
      
      # Stream directly for the user - use either Current.user or connection.current_user
      stream_for user
      
      # Track user presence for potential future features
      appear
      
      # Keep track of subscription time for analytics or rate limiting
      @subscribed_at = Time.zone.now
    end

      # Handles cleanup when the channel is unsubscribed
    # @return [void]
    def unsubscribed
      # Notify presence system the user has gone away
      disappear
      
      # Log disconnect time for analytics if needed
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      if @subscribed_at && user
        Rails.logger.debug { "User #{user.id} disconnected from NotificationChannel after #{Time.zone.now - @subscribed_at} seconds" }
      end
    end
    
    # Track when a user appears (connects)
    # @return [void]
    def appear
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      return unless user # Skip if no user
      
      ActionCable.server.broadcast("presence", { user_id: user.id, event: "online" })
    end
    
    # Track when a user disappears (disconnects)
    # @return [void]
    def disappear
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      return unless user # Skip if no user
      
      ActionCable.server.broadcast("presence", { user_id: user.id, event: "offline" })
    end
    
    # Respond to pings to keep connection alive
    # @return [void]
    def pong
      # Simply respond to client pings to maintain connection health
    end
    
    # Marks a notification as read
    # @param data [Hash] Parameters containing the notification id
    # @option data [String] :id The ID of the notification to mark as read
    # @return [void]
    def mark_as_read(data)
      notification = find_notification(data['id'])
      return reject_unauthorized_action unless notification && can_modify_notification?(notification)
      
      # Use transaction for data consistency
      ActiveRecord::Base.transaction do
        notification.mark_as_read!
        
        # Record when the notification was read for analytics
        notification.update(read_at: Time.zone.now) unless notification.read_at
      end
      
      # Acknowledge successful action to the client
      transmit({ status: "success", action: "mark_as_read", id: notification.id })
      
      # Broadcast the updated unread count to keep all clients in sync
      broadcast_unread_count
    end
    
    # Dismisses a notification
    # @param data [Hash] Parameters containing the notification id
    # @option data [String] :id The ID of the notification to dismiss
    # @return [void]
    def dismiss(data)
      notification = find_notification(data['id'])
      return unless notification && can_modify_notification?(notification)
      
      notification.dismiss!
      
      # Broadcast the updated unread count and dismissal event to keep all clients in sync
      broadcast_to(current_user, { 
        type: 'notification_dismissed', 
        id: notification.id,
        timestamp: Time.zone.now.to_i
      })
      broadcast_unread_count
    end
    
    # Marks a notification as displayed
    # @param data [Hash] Parameters containing the notification id
    # @option data [String] :id The ID of the notification to mark as displayed
    # @return [void]
    def mark_as_displayed(data)
      notification = find_notification(data['id'])
      return unless notification && can_modify_notification?(notification)
      
      notification.mark_as_displayed!
    end
    
    # Request the current unread count from the server
    # This method can be called from the client to get updated counts
    # @return [void]
    def request_count
      # Just call the broadcast_unread_count method to send the current count to the client
      broadcast_unread_count
    end
    
    private
    
    # Finds a notification by ID with enhanced error handling
    # @param id [String] The ID of the notification to find
    # @return [Hub::Notification, nil] The notification if found, nil otherwise
    def find_notification(id)
      Hub::Notification.find_by(id: id)
    rescue ActiveRecord::RecordNotFound
      # Get user ID from either source
      user_id = Current.user&.id || connection.current_user&.id || 'unknown'
      Rails.logger.warn { "User #{user_id} attempted to access non-existent notification: #{id}" }
      nil
    end
    
    # Checks if the current user can modify the notification
    # @param notification [Hub::Notification] The notification to check
    # @return [Boolean] True if the user can modify the notification, false otherwise
    def can_modify_notification?(notification)
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      user && notification.user_id == user.id
    end
    
    # Handles unauthorized actions gracefully
    # @return [void]
    def reject_unauthorized_action
      transmit({ status: "error", message: "Unauthorized action" })
    end
    
    # Rejects the connection with an error message
    # @param message [String] The error message to send
    # @return [void]
    def reject_connection(message)
      # Log the rejection for monitoring
      Rails.logger.warn { "Rejecting NotificationChannel connection: #{message}" }
      
      # Send a rejection message to the client
      transmit({ status: "error", message: message })
      
      # Return false to prevent subscription
      false
    end
    
    # Broadcasts the current user's unread notification count
    # @return [void]
    def broadcast_unread_count
      # Get user from either Current.user or connection.current_user
      user = Current.user || connection.current_user
      # Ensure we have a user
      return unless user
      
      # Use Solid Cache for better performance on counting queries
      # This leverages Rails 8's improved caching capabilities
      unread_count = Rails.cache.fetch("user_#{user.id}_unread_count", expires_in: 5.minutes) do
        Hub::Notification.where(user: user).unread.count
      end
      
      # Broadcast to the specific user's channel with just the unread count
      broadcast_to(user, { 
        type: 'unread_count_updated', 
        count: unread_count,
        timestamp: Time.zone.now.to_i
      })
    end
  end
end
